### PR TITLE
[feat] Triton fused dropout/bias

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?

TODO:
- [x] Implement fused FW
- [x] Implement BW
- [x] Fuse an activation (including BW)
- [x] Add a dedicated layer and hook it up to FusedMLP

 --- Type: torch.float16 --- 
| Units: GB/s                                    |B=8, M=256, K=512   |B=8, M=512, K=1024  |B=4, M=1024, K=1024 |B=2, M=2048, K=2048 |B=2, M=4096, K=4096 |B=1, M=2048, K=12288|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|pytorch - bias: True - fw+bw - act: gelu        |37.9                |60.2                |60.2                |67.4                |75.4                |74.4                |
|triton - bias: True - fw+bw - act: gelu         |45.5                |70.6                |70.6                |80.5                |92.1                |90.5                |

 --- Type: torch.float32 --- 
| Units: GB/s                                    |B=8, M=256, K=512   |B=8, M=512, K=1024  |B=4, M=1024, K=1024 |B=2, M=2048, K=2048 |B=2, M=4096, K=4096 |B=1, M=2048, K=12288|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|pytorch - bias: True - fw+bw - act: gelu        |67.7                |77.5                |77.5                |82.9                |88.3                |87.7                |
|triton - bias: True - fw+bw - act: gelu         |76.6                |106.7               |106.7               |119.8               |130.0               |130.8               |

 --- Type: torch.float16 --- 
| Units: GB/s                                    |B=8, M=256, K=512   |B=8, M=512, K=1024  |B=4, M=1024, K=1024 |B=2, M=2048, K=2048 |B=2, M=4096, K=4096 |B=1, M=2048, K=12288|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|pytorch - bias: False - fw+bw - act: gelu       |54.6                |74.5                |74.1                |82.3                |90.5                |89.4                |
|triton - bias: False - fw+bw - act: gelu        |52.5                |78.0                |78.0                |88.3                |98.8                |97.4                |

 --- Type: torch.float32 --- 
| Units: GB/s                                    |B=8, M=256, K=512   |B=8, M=512, K=1024  |B=4, M=1024, K=1024 |B=2, M=2048, K=2048 |B=2, M=4096, K=4096 |B=1, M=2048, K=12288|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|pytorch - bias: False - fw+bw - act: gelu       |81.1                |94.2                |94.4                |100.4               |105.5               |104.9               |
|triton - bias: False - fw+bw - act: gelu        |91.0                |121.8               |122.3               |134.8               |142.4               |143.6               |

 --- Type: torch.float16 --- 
| Units: GB/s                                    |B=8, M=256, K=512   |B=8, M=512, K=1024  |B=4, M=1024, K=1024 |B=2, M=2048, K=2048 |B=2, M=4096, K=4096 |B=1, M=2048, K=12288|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|pytorch - bias: True - fw - act: gelu           |146.3               |174.3               |174.3               |189.4               |201.0               |199.9               |
|triton - bias: True - fw - act: gelu            |178.1               |292.6               |292.6               |334.4               |369.2               |364.2               |

 --- Type: torch.float32 --- 
| Units: GB/s                                    |B=8, M=256, K=512   |B=8, M=512, K=1024  |B=4, M=1024, K=1024 |B=2, M=2048, K=2048 |B=2, M=4096, K=4096 |B=1, M=2048, K=12288|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|pytorch - bias: True - fw - act: gelu           |210.1               |232.4               |232.4               |245.5               |254.3               |253.4               |
|triton - bias: True - fw - act: gelu            |341.4               |565.0               |565.0               |649.0               |725.2               |717.7               |

 --- Type: torch.float16 --- 
| Units: GB/s                                    |B=8, M=256, K=512   |B=8, M=512, K=1024  |B=4, M=1024, K=1024 |B=2, M=2048, K=2048 |B=2, M=4096, K=4096 |B=1, M=2048, K=12288|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|pytorch - bias: False - fw - act: gelu          |195.0               |252.1               |252.1               |273.1               |288.7               |287.4               |
|triton - bias: False - fw - act: gelu           |178.1               |292.6               |297.9               |334.4               |371.3               |366.8               |

 --- Type: torch.float32 --- 
| Units: GB/s                                    |B=8, M=256, K=512   |B=8, M=512, K=1024  |B=4, M=1024, K=1024 |B=2, M=2048, K=2048 |B=2, M=4096, K=4096 |B=1, M=2048, K=12288|
|------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
|pytorch - bias: False - fw - act: gelu          |282.5               |334.4               |337.8               |356.2               |369.7               |367.5               |
|triton - bias: False - fw - act: gelu           |341.3               |574.9               |574.9               |662.0               |736.4               |728.2               |
## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by blefaudeux** [comment]
I've a lingering bug in the gradient computation that the unit test catches, fixing that and re-submitting. Speed looks good, checked on V100 and a laptop

**Comment by blefaudeux** [comment]
I've a lingering bug in the gradient computation that the unit test catches, fixing that and re-submitting. Speed looks good, checked on V100 and a laptop

File: BENCHMARKS.md [hunk 1]
```diff
@@ -42,52 +42,65 @@ Some examples, generated with `python3 xformers/benchmarks/benchmark_encoder.py
 You can reproduce these numbers locally by running `python3 xformers/benchmarks/benchmark_triton_softmax.py`. The units are GB/s. These results are for a nVidia V100, Triton 1.1 and PyTorch 1.9.
 Note that in the Triton case the slowdowns at extreme sizes are because of register spilling, A100s get much better performance.
 
-![Softmax throughput in fp16 - inference](docs/plots/Softmax_Bandwidth_FW_fp16.png)
+![Softmax throughput in fp16 - inference](docs/plots/fused_softmax/Softmax_Bandwidth_FW_fp16.png)
```
Comment (blefaudeux) [inline BENCHMARKS.md:45]: moving all the plots to subfolders to clean that up a little

**Comment by blefaudeux** [comment]
Full training checked on the microViT example, all good (slightly better results for some reason, 15-20% better speed)

**Comment by blefaudeux** [comment]
Apples to apples comparison:
Before
https://app.circleci.com/pipelines/github/facebookresearch/xformers/426/workflows/a578f508-7576-4655-9ca2-38c72e03917a/jobs/869/parallel-runs/0/steps/0-111

After
https://app.circleci.com/pipelines/github/facebookresearch/xformers/433/workflows/f08860f8-cea8-47b7-9050-f9d7bf846ece/jobs/883/parallel-runs/0/steps/0-111

We're now even a tiny bit faster with the layernorm change which landed in between

**Comment by blefaudeux** [comment]
ping review @jieru-hu @dianaml0 , if you don't mind ? This should be ready

**Review by jieru-hu — APPROVED** [review]

File: xformers/benchmarks/benchmark_triton_dropout.py [hunk 2]
```diff
@@ -69,17 +91,21 @@ def triton_step(x):
...
+                bias, "+BW" if backward else "", dtype, activation
+            ),
             units="GB/s",
             dash_key="pytorch",
         )
 
 
-for bw in [False, True]:
-    bench_dropout(bw)
+for activation in [Activation.GeLU, None]:
```
Comment (dianaml0) [inline xformers/benchmarks/benchmark_triton_dropout.py:108]: Should others be benchmarked as well or its pretty similar for all of them?

**Review by dianaml0 — APPROVED** [review]
Nice improvement in GB/s! :)

File: xformers/benchmarks/benchmark_triton_dropout.py (hunk 2 above)
Comment (blefaudeux) [reply xformers/benchmarks/benchmark_triton_dropout.py:108]: Gelu is kind of the worst case for activations and Triton (it uses the real exp function and not the fastmath equivalent), and "None" negates the fusion (well, we fuse with nothingness :D), so these are kind of hard goals to test truthfully. It's a little slow to test with everything, but could be extended on a ad-hoc basis if people would like ?

Comment (dianaml0) [reply xformers/benchmarks/benchmark_triton_dropout.py:108]: Ah okay, that makes sense! Guess its pretty straightforward to modify and add an activation if someone needs to
